music21/musicxml/xmlToM21.py

```diff
@@ -1586,8 +1587,30 @@ def _adjustMidiData(mc):
 
         seta = _setAttributeFromTagText
 
-        # print(ET.tostring(mxScorePart, encoding='unicode'))
-        i = instrument.Instrument()
+        # for now, just get first midi instrument
+        # TODO: get all
+        # TODO: midi-device
+        # TODO: midi-name
+        # TODO: midi-bank transform=_adjustMidiData
+        # TODO: midi-unpitched
+        # TODO: midi-volume
+        # TODO: pan
+        # TODO: elevation
+        # TODO: store id attribute somewhere
+        mxMIDIInstrument = mxScorePart.find('midi-instrument')
+        i: Optional[instrument.Instrument] = None
+        if mxMIDIInstrument is not None:
+            mxMidiProgram = mxMIDIInstrument.find('midi-program')
+            if textStripValid(mxMidiProgram):
+                try:
+                    i = instrument.instrumentFromMidiProgram(_adjustMidiData(mxMidiProgram.text))
+                except instrument.InstrumentException as ie:
+                    warnings.warn(MusicXMLWarning(ie))
+                    i = instrument.Instrument()
+                seta(i, mxMIDIInstrument, 'midi-channel', transform=_adjustMidiData)
+        if i is None:
+            i = instrument.Instrument()
```

Comment:
Seeing all the Sopranos/Altos/etc. with `music21.instrument.Instrument` tags makes me think that something's missing here.  Some possibilities:

We should be reclassifying based on the Instrument name in preference to the midi-program, if possible.  That would solve the Sopranos, etc. (we might need to create some new Instrument objects for Soprano, Alto, Tenor, Baritone, Bass though).  Or we can special-case midi-program 0, unless piano is in the text OR max-staves >= 2.   MIDI 0 is often used just to mean I didn't bother to specify the instrument.

For me, seeing a mislabeled "Piano" on things that are clearly not pianos is worse than a generic Instrument. tag.

Does MIDI-Program ever get set in case no instrument can be found?  I think there are some midi-programs that don't map to one of our instruments [on purpose] such as FX-Goblin or TremoloStrings, and we should preserve the number, even if we can only get i = instrument.Instrument()

Reply:
> For me, seeing a mislabeled "Piano" on things that are clearly not pianos is worse than a generic Instrument. tag.

I was thinking we could iterate on this. Just wanted to get a patch going for discussion to get the implementation agreed on. Yes, we could special case MIDI number 0 (1 in mxml).

Our fromString methods handle all the easy cases, but it seems more discrete and reliable to privilege the MIDI data if we can, but yes, absolutely special case 0 to go to the instrument name instead.

In terms of being worse than a generic instrument, it _would_ get worse once we do more `if isinstance(inst, instrument.UnpitchedPercussion)`, etc., which I think we already do (or should be doing) for MIDI export.

Reply:
> Does MIDI-Program ever get set in case no instrument can be found? I think there are some midi-programs that don't map to one of our instruments [on purpose] such as FX-Goblin or TremoloStrings, and we should preserve the number, even if we can only get i = instrument.Instrument()

They all have fallbacks like `Sampler` #757 and even the generic fallback would make `i` not None; `if i is None` is just paranoia to catch if `mxMidiInstrument` is None (encoding error?) or if `textStripValid` rejected some garbage -- clearly a comment is in order!

Reply:
> a fromString() operation on the name, although this could be added later (for instance to change Piano to Soprano when a file used piano sounds rather than Choir Aahs)